packages/react-dom-bindings/src/client/ReactFiberConfigDOM.js

```diff
@@ -4522,6 +4522,79 @@ function getScriptSelectorFromKey(key: string): string {
   return 'script[async]' + key;
 }
 
+// for the sake of performance we want to use insertRule
+// if that fails, fall back to appending textContent (which is slow)
+function appendStyleRule(style: HTMLStyleElement, cssText: string) {
+  try {
+    if (style.sheet)
+      style.sheet.insertRule(cssText, style.sheet.cssRules.length);
+    else style.textContent += cssText;
+  } catch (e) {
+    style.textContent += cssText;
+  }
+}
```

Comment:
Please take my feedback with a grain of salt. I'm not a React team member - but we were chatting with @gnoff about most of this already.

I think this change should be reverted (for now). I totally agree this is something that will have to be implemented in some capacity at some point but there are open design questions around this. 

It feels like, at the moment, this breaks the core assumption that a single resource is represented by a single element. OTOH, not many tests were affected by this change and that makes me doubt a little bit in my interpretation of the change.

All in all, I'd focus first on caching insertion points for each precedence to avoid expensive and redundant querying.